ledger/participant-integration-api/src/main/scala/platform/store/dao/EventProjectionProperties.scala

```diff
@@ -0,0 +1,109 @@
+
+/**  This class encapsulates the logic of how contract arguments and interface views are
+  *  being projected to the consumer based on the filter criteria and the relation between
+  *  interfaces and templates implementing them.
+  *
+  * @param verbose enriching in verbose mode
+  * @param populateContractArgument populate contract_argument, and contract_key. If templateId set is empty: populate.
+  * @param populateInterfaceView populate interface_views. The Map of templates to interfaces,
+  *                              and the set of implementor templates cannot be empty.
+  */
+final case class EventProjectionProperties private[dao] (
+    verbose: Boolean,
+    // Map(eventWitnessParty, Set(templateId))
+    populateContractArgument: Map[String, Set[Identifier]] = Map.empty,
+    // Map(eventWitnessParty, Map(templateId -> Set(interfaceId)))
+    populateInterfaceView: Map[String, Map[Identifier, Set[Identifier]]] = Map.empty,
+) {
+  def render(witnesses: Seq[String], templateId: Identifier): RenderResult = {
+    val renderContractArguments = witnesses.view
+      .flatMap(populateContractArgument.get)
+      .exists(templates => templates.isEmpty || templates(templateId))
+
+    val renderInterfaces: Set[Identifier] =
+      (for {
+        witness <- witnesses.iterator
+        templateToInterfaceMap <- populateInterfaceView.get(witness).iterator
+        interfaces <- templateToInterfaceMap.getOrElse(templateId, Set.empty[Identifier])
+      } yield interfaces).toSet
+
+    RenderResult(renderContractArguments, renderInterfaces)
+  }
+
+}
+
+object EventProjectionProperties {
+
+  case class RenderResult(
+      contractArguments: Boolean,
+      interfaces: Set[Identifier],
+  )
+
+  /** @param domainTransactionFilter Transaction filter as defined by the consumer of the API.
+    * @param verbose                 enriching in verbose mode
+    * @param interfaceImplementedBy  The relation between an interface id and template id.
+    *                                If template has no relation to the interface,
+    *                                an empty Set must be returned.
+    */
+  def apply(
+      domainTransactionFilter: domain.TransactionFilter,
+      verbose: Boolean,
+      interfaceImplementedBy: Identifier => Set[Identifier],
+  ): EventProjectionProperties =
+    EventProjectionProperties(
+      verbose = verbose,
+      populateContractArgument = populateContractArgument(domainTransactionFilter),
+      populateInterfaceView = populateInterfaceView(domainTransactionFilter, interfaceImplementedBy),
+    )
+
+  private def populateContractArgument(
+      domainTransactionFilter: domain.TransactionFilter
+  ): Map[String, Set[Identifier]] = {
+
+    val wildcardFilters = domainTransactionFilter.filtersByParty.iterator
+      .collect {
+        case (party, Filters(None)) => party
+        case (party, Filters(Some(empty)))
+            if empty.templateIds.isEmpty && empty.interfaceFilters.isEmpty =>
```

Comment:
This looks like a partition but not full (e.g. checks empty.interfaceFilters.isEmpty but doesn't use it in the `templateFilters`). Semantically it is possible to have `interfaceFilters` nonEmpty so this wildcard entry would be skipped even if `templateIds` is empty. Not sure whether this is a problem (maybe validation will always enforce one empty and one not) or whether it is going to be. wdyt?

Reply:
Furthermore, if we can make this partition full, I think this method can be simplified